File: bridge/python/src/destack/protocol/codec.py
from __future__ import annotations

from collections.abc import Sequence

from destack._generated.protocol.envelope import (
    ProtocolMessage,
    decode_protocol_message,
    encode_protocol_message,
)
from destack.protocol.serde import decode_value, encode_value

FRAME_MAGIC = bytes([0x44, 0x53, 0x57, 0x50])
FRAME_VERSION = 1
FRAME_HEADER_SIZE = 12


class ProtocolCodecError(Exception):
    """Error thrown while encoding or decoding workspace protocol frames."""


def encode_message(message: ProtocolMessage) -> bytes:
    """Encode one protocol message payload."""

    return encode_value(lambda writer: encode_protocol_message(writer, message))


def decode_message(payload: bytes | bytearray | Sequence[int]) -> ProtocolMessage:
    """Decode one protocol message payload."""

    return decode_value(payload, decode_protocol_message)


def encode_frame(message: ProtocolMessage) -> bytes:
    """Encode one protocol message frame."""

    payload = encode_message(message)

    return encode_payload_frame(payload)


def decode_frame(frame: bytes | bytearray | Sequence[int]) -> ProtocolMessage:
    """Decode one protocol message frame."""

    payload = decode_payload_frame(frame)

    return decode_message(payload)


def encode_payload_frame(payload: bytes | bytearray | Sequence[int]) -> bytes:
    """Encode one protocol payload frame."""

    payload = bytes(payload)
    header = encode_header(len(payload))

    return header + payload


def decode_payload_frame(frame: bytes | bytearray | Sequence[int]) -> bytes:
    """Decode one protocol payload frame."""

    frame = bytes(frame)
    payload_len = decode_header(frame)
    end = FRAME_HEADER_SIZE + payload_len
    if len(frame) < end:
        raise ProtocolCodecError("unexpected end of frame payload")

    return frame[FRAME_HEADER_SIZE:end]


def encode_header(payload_len: int) -> bytes:
    """Encode one frame header."""

    if not isinstance(payload_len, int) or payload_len < 0 or payload_len > 0xFFFFFFFF:
        raise ProtocolCodecError(f"invalid payload length: {payload_len}")

    return (
        FRAME_MAGIC
        + FRAME_VERSION.to_bytes(2, "little")
        + (0).to_bytes(2, "little")
        + payload_len.to_bytes(4, "little")
    )


def decode_header(frame: bytes) -> int:
    """Decode one frame header and return its payload length."""

    if len(frame) < FRAME_HEADER_SIZE:
        raise ProtocolCodecError("unexpected end of frame header")

    magic = frame[:4]
    version = int.from_bytes(frame[4:6], "little")
    payload_len = int.from_bytes(frame[8:12], "little")
    if magic != FRAME_MAGIC:
        raise ProtocolCodecError("invalid frame magic")
    if version != FRAME_VERSION:
        raise ProtocolCodecError(f"unsupported frame version: {version}")

    return payload_len

File: bridge/typescript/src/protocol/codec.ts
import type { ProtocolMessage } from "../_generated/protocol/envelope.js";
import { decodeProtocolMessage, encodeProtocolMessage } from "../_generated/protocol/envelope.js";
import { decodeValue, encodeValue } from "./serde.js";

const FRAME_MAGIC = [0x44, 0x53, 0x57, 0x50] as const;
const FRAME_VERSION = 1;
const FRAME_HEADER_SIZE = 12;

/** Error thrown while encoding or decoding workspace protocol frames. */
export class ProtocolCodecError extends Error {
    /** Create one protocol codec error. */
    constructor(message: string) {
        super(message);
        this.name = "ProtocolCodecError";
    }
}

/** Header fields for one workspace protocol frame. */
export type FrameHeader = {
    /** Magic bytes for protocol frames. */
    readonly magic: readonly [number, number, number, number];
    /** Frame format version. */
    readonly version: number;
    /** Reserved flags. */
    readonly flags: number;
    /** Payload length in bytes. */
    readonly payloadLength: number;
};

/** Encode one protocol message payload. */
export function encodeMessage(message: ProtocolMessage): Uint8Array {
    return encodeValue((writer) => {
        encodeProtocolMessage(writer, message);
    });
}

/** Decode one protocol message payload. */
export function decodeMessage(payload: Uint8Array | readonly number[]): ProtocolMessage {
    return decodeValue(payload, decodeProtocolMessage);
}

/** Encode one protocol message frame. */
export function encodeFrame(message: ProtocolMessage): Uint8Array {
    const payload = encodeMessage(message);

    return encodePayloadFrame(payload);
}

/** Decode one protocol message frame. */
export function decodeFrame(frame: Uint8Array | readonly number[]): ProtocolMessage {
    const payload = decodePayloadFrame(frame);

    return decodeMessage(payload);
}

/** Encode one protocol payload frame. */
export function encodePayloadFrame(payload: Uint8Array | readonly number[]): Uint8Array {
    const bytes = payload instanceof Uint8Array ? payload : new Uint8Array(payload);
    const header = encodeHeader(bytes.length);
    const frame = new Uint8Array(FRAME_HEADER_SIZE + bytes.length);

    frame.set(header, 0);
    frame.set(bytes, FRAME_HEADER_SIZE);

    return frame;
}

/** Decode one protocol payload frame. */
export function decodePayloadFrame(frame: Uint8Array | readonly number[]): Uint8Array {
    const bytes = frame instanceof Uint8Array ? frame : new Uint8Array(frame);
    const header = decodeHeader(bytes);
    const end = FRAME_HEADER_SIZE + header.payloadLength;

    if (bytes.length < end) {
        throw new ProtocolCodecError("unexpected end of frame payload");
    }

    return bytes.slice(FRAME_HEADER_SIZE, end);
}

/** Encode one frame header. */
export function encodeHeader(payloadLength: number): Uint8Array {
    if (!Number.isInteger(payloadLength) || payloadLength < 0 || payloadLength > 0xffffffff) {
        throw new ProtocolCodecError(`invalid payload length: ${payloadLength}`);
    }

    const bytes = new Uint8Array(FRAME_HEADER_SIZE);
    const view = new DataView(bytes.buffer);

    bytes.set(FRAME_MAGIC, 0);
    view.setUint16(4, FRAME_VERSION, true);
    view.setUint16(6, 0, true);
    view.setUint32(8, payloadLength, true);

    return bytes;
}

/** Decode one frame header. */
export function decodeHeader(bytes: Uint8Array): FrameHeader {
    if (bytes.length < FRAME_HEADER_SIZE) {
        throw new ProtocolCodecError("unexpected end of frame header");
    }

    const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    const magic = [bytes[0], bytes[1], bytes[2], bytes[3]] as const;
    const version = view.getUint16(4, true);
    const flags = view.getUint16(6, true);
    const payloadLength = view.getUint32(8, true);

    if (!magic.every((byte, index) => byte === FRAME_MAGIC[index])) {
        throw new ProtocolCodecError("invalid frame magic");
    }

    if (version !== FRAME_VERSION) {
        throw new ProtocolCodecError(`unsupported frame version: ${version}`);
    }

    return {
        magic,
        version,
        flags,
        payloadLength,
    };
}
